examples/credit-risk-end-to-end/02_Deploying_the_Feature_Store.ipynb

```diff
@@ -0,0 +1,825 @@
+    "- Entity: primary key fields which can be used for joining data\n",
+    "- FeatureView: collections of features from a data source\n",
+    "- FeatureService: collection of FeatureViews\n",
+    "\n",
+    "For more information on these, see the [Concepts](https://docs.feast.dev/getting-started/concepts) section of the Feast documentation."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "d3e8fd80-0bee-463c-b3fb-bd0d1ee83a9c",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Writing Feature_Store/feature_definitions.py\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%writefile Feature_Store/feature_definitions.py\n",
+    "\n",
+    "# Imports\n",
+    "import os\n",
+    "from pathlib import Path\n",
+    "from feast import (\n",
+    "    FileSource,\n",
+    "    Entity,\n",
+    "    FeatureView,\n",
+    "    Field,\n",
+    "    FeatureService\n",
+    ")\n",
+    "from feast.types import Float32\n",
+    "from feast.data_format import ParquetFormat\n",
+    "\n",
+    "CURRENT_DIR = os.path.abspath(os.curdir)\n",
+    "\n",
+    "# Data Sources\n",
+    "# A data source tells Feast where the data lives\n",
+    "data_a = FileSource(\n",
+    "    file_format=ParquetFormat(),\n",
+    "    path=Path(CURRENT_DIR,\"data/train_a.parquet\").as_uri()\n",
+    ")\n",
+    "data_b = FileSource(\n",
+    "    file_format=ParquetFormat(),\n",
+    "    path=Path(CURRENT_DIR,\"data/train_b.parquet\").as_uri()\n",
+    ")\n",
+    "data_test = FileSource(\n",
+    "    file_format=ParquetFormat(),\n",
+    "    path=Path(CURRENT_DIR,\"data/test.parquet\").as_uri()\n",
+    ")   \n",
+    "\n",
+    "# Entity\n",
+    "# An entity tells Feast the column it can use to join tables\n",
+    "loan_id = Entity(\n",
+    "    name = \"loan_id\",\n",
+    "    join_keys = [\"ID\"]\n",
+    ")\n",
+    "\n",
+    "# Feature views\n",
+    "# A feature view is how Feast groups features\n",
+    "features_a = FeatureView(\n",
+    "    name=\"data_a\",\n",
+    "    entities=[loan_id],\n",
+    "    schema=[\n",
+    "        Field(name=\"duration\", dtype=Float32),\n",
+    "        Field(name=\"credit_amount\", dtype=Float32),\n",
+    "        Field(name=\"installment_commitment\", dtype=Float32),\n",
+    "        Field(name=\"checking_status_ord\", dtype=Float32)\n",
+    "    ],\n",
+    "    source=data_a\n",
+    ")\n",
+    "features_b = FeatureView(\n",
+    "    name=\"data_b\",\n",
+    "    entities=[loan_id],\n",
+    "    schema=[\n",
+    "        Field(name=\"residence_since\", dtype=Float32),\n",
+    "        Field(name=\"age\", dtype=Float32),\n",
+    "        Field(name=\"existing_credits\", dtype=Float32),\n",
+    "        Field(name=\"num_dependents\", dtype=Float32),\n",
+    "        Field(name=\"housing_ord\", dtype=Float32)\n",
+    "    ],\n",
+    "    source=data_b\n",
+    ")\n",
+    "features_test = FeatureView(\n",
```

Comment:
Wouldn't it be more realistic to do train/test split after `get_historical_features`? a separate FV for a test dataset might be more misleading/confusing for someone trying to learn feast.

Reply:
+1

Reply:
I would say that's not the pattern we want users to adopt

Reply:
Here's the issue I'm seeing. A test set should be split out before EDA and feature engineering occurs to prevent data snooping bias (see [Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow, 2nd Ed](https://www.amazon.com/Hands-Machine-Learning-Scikit-Learn-TensorFlow/dp/1492032646) page 51). This is actually pretty important. (I didn't make up data snooping bias -- it's an industry term, just in case anyone was wondering.)

However, to `get_historical_features`, you have presumably already done EDA and feature engineering on the underlying data so that you could set up your FeatureViews. Splitting out test data after that would not be recommended.

For this example, I could change the 'test' set to a 'validation' set (or a cross-validation paradigm), which would be split out after feature engineering. (I think this is in line with what you are thinking). In this case, I would just leave the true test set concept out.

We probably need to come up with a pattern for how Feast deals with a true test set scenario. I'll ping in the Feast chat about this.

Reply:
Oh, I missed the encoder part in the first notebook. I'm no data scientist, but fortunately I have read that page :laughing:. tbh, I dislike the example even more now lol.

Correct me if I'm wrong, but I think there's a difference between the kind of feature engineering that happens before feature store and the one that happens after. Most of the features a data scientist would find in an existing feature store while starting to work a problem will be either "raw" descriptive features or features that have been previously crafted by others for other use cases. true, they are probably a product of an EDA, but done by someone else a long time ago. Also, they are never sitting right next to one another like in this example, so it would be physically impossible to do your model-specific EDA before `get_historical_features`. You would realistically call `get_historical_features` on all entities, do a train/test split and a further EDA would come after that.

Another problem with applying OrdinalEncoder in this order is that although you're correctly fitting it only on the train dataset and applying on the test, the example then goes on to store calculated feature values in feast effectively discarding information stored inside fitted OrdinalEncoder object (in case of OrdinalEncoder, mapping between strings and ints). Practically speaking, even if you find the trained model satisfactory, you won't be able to deploy it as you won't be able to apply an encoder to new values that will come to you during inference.

I guess what I'm trying to say is that not all kinds of feature engineering need to feed into the feature store and encoders fitted to a model-specific training sets are a clear example. They should come after `get_historical_features` and more importantly they should be bundled into the resulting model if the goal is for the end product to be deployable (think of something like a sklearn pipeline here).

Reply:
> Correct me if I'm wrong, but I think there's a difference between the kind of feature engineering that happens before feature store and the one that happens after.

Right, notebook 01 starts in the context of "feature engineering before feature store". 

> true, they are probably a product of an EDA, but done by someone else a long time ago.

Unfortunately, bias is transitive--this is what makes it really hard to fix generalization error. 

> it would be physically impossible to do your model-specific EDA before get_historical_features.

This is routine data science -- munging data together. It's done all the time outside of Feast.

> you won't be able to deploy it as you won't be able to apply an encoder to new values that will come to you during inference.

Actually, this is pretty common. With Sklearn, you pickle the encoder with the model in an [inference pipeline](https://scikit-learn.org/1.5/modules/generated/sklearn.pipeline.Pipeline.html), on which you call "predict" just like a simple model.

Now, if I'm understanding right, you would like to see the example flow like this:

- Feast data sources point to raw data
- "training" data pulled (get_historical_features), split, and EDA and final feature engineering completed. Then model trained and validated.
- serving data (get_online_features) fed to inference pipeline to get prediction.

I can refactor the example in this way. Let me know what you thin @tokoko .

Reply:
I was just noticing that none of the other Feast examples (in the repo, or in Feast.dev parent) even mention a validation dataset. The one's that have models do model.fit, and then go straight to serving. 

As we've just discussed, it's not trivial for a user to determine where validation and test sets come in when using Feast. Just getting a validation set in this example, will be a big win for the examples section.

Reply:
yeah, that's right. at least that's the flow that feels most intuitive for me.

- split happens after get_historical_features.
- final part of feature engineering (stuff like OrdinalEncoder, StandardScaler and so on) are applied after split.

Reply:
> Correct me if I'm wrong, but I think there's a difference between the kind of feature engineering that happens before feature store and the one that happens after. 

In the short term, I think this is true. As I mentioned today, I do believe it'd be ideal to get Feast to a state where all transformations happen in Feast and model inference can operate on a DGBMatrix, Torch Tensor, or numpy matrix. It turns out that a lot of skew can happen during matrix serialization and getting that centralized in Feast is a worth endeavor.

Reply:
> Actually, this is pretty common. With Sklearn, you pickle the encoder with the model in an inference pipeline, on which you call "predict" just like a simple model.

Yes, this entire pipeline can introduce latency for lots of production use cases and miscellaneous skew issues. It's why I said I think long term we should work to move these things in Feast.

Reply:
> Since the validation/test is how a data scientist makes decisions and assesses model performance, its really important--and features play a role here.

I agree it's important. It should also be a simple post processing step after `get_online_features`. People cut their training/test/validation sets differently depending upon their business requirements. That's, of course, something we could implement during some sort of Data Version Control process but we'd need to build that.

Reply:
Alright, this has been good feedback. @tokoko I reordered the steps according to our discussion above, and I renamed the 'test' set to a 'validate' set (not a big issue, but just to avoid any confusion). I was able to simplify the notebook flow a bit as well.

Reply:
Just checking back. I believe this is the only outstanding question on this PR. With the changes I made, is this ready to be merged?

Reply:
Checking one more time. I believe this question has been resolved.